include/odp/api/spec/timer.h

```diff
@@ -294,73 +296,86 @@ odp_timer_t odp_timer_alloc(odp_timer_pool_t tpid,
+ *                            the current time.
+ * @retval ODP_TIMER_TOOLATE  Failure. Expiration time is too far from
+ *                            the current time.
+ * @retval ODP_TIMER_NOEVENT  Failure. Set operation: no event provided.
+ *                            Reset operation: timer already expired.
  *
- * @retval ODP_TIMER_SUCCESS Operation succeeded
- * @retval ODP_TIMER_TOOEARLY Operation failed because expiration tick too
- * early
- * @retval ODP_TIMER_TOOLATE Operation failed because expiration tick too
- * late
- * @retval ODP_TIMER_NOEVENT Operation failed because timeout event not
- * specified in odp_timer_set call and not present in timer
+ * @see odp_timer_set_rel(), odp_timer_alloc(), odp_timer_cancel()
  */
-int odp_timer_set_abs(odp_timer_t tim,
-		      uint64_t abs_tck,
+int odp_timer_set_abs(odp_timer_t timer, uint64_t abs_tick,
 		      odp_event_t *tmo_ev);
 
 /**
- * Set a timer with a relative expiration time and user-provided event.
+ * Set (or reset) a timer with relative expiration time
  *
- * Set (arm) the timer to expire at a relative future time.
+ * Like odp_timer_set_abs(), but the expiration time is relative to the current
+ * time: expiration tick = odp_timer_current_tick() + 'rel_tick'.
  *
- * @param tim      Timer
- * @param rel_tck  Expiration time in timer ticks relative to current time of
- *		   the timer pool the timer belongs to
- * @param[in,out] tmo_ev  Reference to an event variable that points to
- * timeout event or NULL to reuse the existing timeout event. Any existing
- * timeout event that is replaced by a successful set operation will be
- * returned here.
+ * @param         timer    Timer
+ * @param         rel_tick Expiration time relative to current time of
+ *                         the timer pool in timer ticks
+ * @param[in,out] tmo_ev   Pointer to an event handle. The event is enqueued
+ *                         when the timer expires. Use NULL when resetting the
+ *                         timer without changing the event. When resetting the
+ *                         timer with a new event, a successful operation
+ *                         outputs the old event here.
  *
- * @retval ODP_TIMER_SUCCESS Operation succeeded
- * @retval ODP_TIMER_TOOEARLY Operation failed because expiration tick too
- * early
- * @retval ODP_TIMER_TOOLATE Operation failed because expiration tick too
- * late
- * @retval ODP_TIMER_NOEVENT Operation failed because timeout event not
- * specified in call and not present in timer
+ * @retval ODP_TIMER_SUCCESS  Success
+ * @retval ODP_TIMER_TOOEARLY Failure. Expiration time is too near to
+ *                            the current time.
+ * @retval ODP_TIMER_TOOLATE  Failure. Expiration time is too far from
+ *                            the current time.
+ * @retval ODP_TIMER_NOEVENT  Failure. Set operation: no event provided.
+ *                            Reset operation: timer already expired.
```

Comment:
I'm not sure I understand the `ODP_TIMER_NOEVENT` case with Reset operation. In this case user gives a "valid" `rel_tick` but during the set/reset operation timer fires and therefore there was no event to "reset"?
Wouldn't that normally mean that platform could simply fire the new event and old event won't be available?
If "inserting" and event into the timer would mean software blocking timer HW from working/ticking, it means that SW will have to take care of any timers that have fired while it processes (i.e. our platform goes through a current bucket and manually fires all events).

That would be a "success" case since the old event is already pushed on a queue.

If that is a failure case then is my assumption correct that implementation should NOT arm the timer?

Reply:
Yes. In reset case, functions arguments were OK (expiration tick value and event), but reset is not possible as the timer cannot be modified any more (has fired already or will fire soon). Reset operation failed, but timer has fired (will fire soon) with the previous settings. The "old" event will be received normally through the queue and the new event is not consumed. So, implementation does not have to do anything extra if it was too late to reset HW timer, just return this failure and the old event is processed normally. Timeout can be marked stale, if there's no danger of race condition (should not write to a memory location that is not any more owned by the implementation).

Reply:
Then perhaps it would be a bit clearer to say "timer expired during the call".

Reply:
The timer may have expired already, or it expires soon (e.g. 1us) enough that it cannot be modified anymore. So, it may not exactly expire during the call...

Reply:
Fair enough. Then please change this to "Too late to modify the timer".

Reply:
Corrected in v4. Also I added this to function description: "A failed reset operation does not modify the timer."